### ISSUE TITLE
should support font fallback

### ISSUE THREAD

**Issue by quininer** [jwilm/alacritty#45]
<img width="763" alt="screenshot_20170103_142910" src="https://cloud.githubusercontent.com/assets/6286761/21601164/c07121e0-d1c1-11e6-8079-1a33321cb4d8.png">

* both konsole and alacritty use Noto Mono.

**Comment by jwilm** [comment]
Fallback fonts are a high priority

**Comment by quininer** [comment]
I found alacritty now can correctly show the chinese char, the issue may be close?

**Comment by IN3D** [comment]
Hi there! I'm still not seeing non-latin characters rendering. The machine I'm on is MacOS Sierra 10.12.3. Alacritty is on the left, iTerm2 on the right. (Apologize in advance for the probably terrible Google Translate powered translations.)

<img width="1260" alt="screen shot 2017-03-23 at 11 56 15 am" src="https://cloud.githubusercontent.com/assets/5441516/24257339/fcc77b02-0fc0-11e7-8d25-c68aa138be67.png">

What I think is the relevant part of my `.alacritty.yml` is:
```yaml
dpi:
  x: 96.0
  y: 96.0

tabspaces: 4

draw_bold_text_with_bright_colors: true

font:
  normal:
    family: "Noto Mono"
    style: "Regular"
  size: 14.0
  offset:
    x: 0.0
    y: 0.0
```

I'm curious what could be allowing the characters to render on one machine, but not the other. I hope this information helps!

**Comment by cpick** [comment]
If anyone is still having an issue on Linux it would be great if they could please test again with my cpick:font-fallback branch from #456 

It will *not* work (yet) on macOS, but if it helps people on Linux that might be a good reason to look into porting the work.

**Comment by DivineGod** [comment]
Has anyone looked at what might be missing to get mac support for fallback fonts?

My custom `fish` prompt uses some unicode characters which it seems alacritty can't find.

characters I am trying to use, which work fine in iTerm2: `ᚦ`,  `ᛝ`, `ᚦ`, `ᚱ`, `ᛄ`

**Comment by jwilm** [comment]
@DivineGod I've looked into font fallback on macOS, and it appears we can use the cascade lists provided by CoreText. From [the docs],

> Core Text font references provide a sophisticated, automatic font-substitution mechanism called font cascading, which picks an appropriate font to substitute for a missing font while taking font traits into account. Font cascading is based on cascade lists, which are arrays of ordered font descriptors. There is a system default cascade list (which is polymorphic, based on the user's language setting and current font) and a font cascade list that is specified at font creation time. Using the information in the font descriptors, the cascading mechanism can match fonts according to style as well as matching characters. The CTFontCreateForString function uses cascade lists to pick an appropriate font to encode a given string. To specify and retrieve font cascade lists, use the kCTFontCascadeListAttribute property.

We would be accessing that `kCTFontCascadeListAttribute` and looking up the font ourselves. The excerpt recommends `CTFontCreateForString` to have it pick for you, but creating a string every time we need to fallback would slow things down.

As with most things, the blocker is time :wink:.

[the docs]: https://developer.apple.com/library/content/documentation/StringsTextFonts/Conceptual/CoreText_Programming/Overview/Overview.html

**Comment by DivineGod** [comment]
@jwilm ah, I was looking through apple dev docs but couldn't find it myself. I haven't touched any parts of apple specific code before, but I'd be happy to give it a go.

Agree on the part of trying to stay away from `CTFrontCreateFromString`, even with caching the result for each glyph it'll probably be some overhead.

**Comment by algesten** [comment]
Looking at this I suspect we want to refactor @cpick #456 above to yank the caching aspect up to the font lib level since this will be something both ft/darwin would share.

Also I wonder whether `HashMap` really is the most effective data structure for such as cache, since I expect these fallback lists to be < 10 for the vast majority of language platforms.

Furthermore on the Darwin side, I think using apple's built in default list per language [CTFontCopyDefaultCascadeListForLanguages](http://stackoverflow.com/questions/29069362/how-does-apples-text-rendering-draw-a-glyph-that-a-font-doesnt-have) would be good. But is there something similar on the ft side?

**Comment by algesten** [comment]
Before attempting any further work, I did an investigation to see what [fallback lists Apple actually use](https://gist.github.com/algesten/203b49c2ef98063a9cf2ea554187f57a). It seems they (at the moment), only have three lists, one for japanese, one for korean and one for everything else.

Rust `core-text` doesn't map up `CTFontCopyDefaultCascadeListForLanguages`, so I suggest we make a config option(s) to provide these lists, and initially make hardcoded reasonable defaults. 

Some further investigation is needed to see if the lists differ for Bold/Italic.

Also I assume the major Linux distros have similar fallbacks, so some further work is needed to work out what they are.

**Comment by madnight** [comment]
In many applications like urxvt or polybar its possible to define multiple fonts and in case of a mismatch the next fallback option is taken, that seems to be a simple and os independent solution.

**Comment by algesten** [comment]
@madnight I suspected as much. The question is whether alacritty should provide some reasonable defaults out of the box or not. If you omit the font config in `alacritty.yml` you already get [these defaults](https://github.com/jwilm/alacritty/blob/master/src/config.rs#L1307-L1335). So the question is whether to go further than that and hardcode default fallbacks too.

The config format could be just changing the font definition to arrays of objects.

```yaml
font:
  normal:
    - family: "Source Code Pro"
      style: Light
    - family: "Menlo"
    - family: "Monaco"
```

**Comment by algesten** [comment]
WIP https://github.com/algesten/alacritty/commit/43725bab0d0365392b03528e09caf5cca6ce0a4c

It's quite close to work for macOS now. The idea is to change so that `FontKey` represents a `FontDescList` (instead of a `FontDesc`) so that `GlyphKey` is tied to a list of fallback fonts. `Rasterizer::get_glyph(&GlyphKey)` is changed to try each font in the font list until it finds a glyph and rasterize that. This means `GlyphCache` is left unchanged so there should be minimal speed decrease.

I can see it trying each font in turn, but for some reason I don't get a `glyph_index`. Debugging continues tomorrow.

My `alacritty.yml` that loads all the apple defaults is:

```yaml
# Font configuration (changes require restart)
font:
  # The normal (roman) font face to use.
  normal:
    - family: "Menlo" # should be "Menlo" or something on macOS.
      style: Regular  # Style can be specified to pick a specific face.
    - family: "Monaco"
    - family: "Lucida Grande"
    - family: "Courier New"
    - family: "Ayuthaya"
    - family: "Kailasa"
    - family: "PingFang SC"
    - family: "PingFang TC"
    - family: "Hiragino Sans"
      style: W3
    - family: "Hiragino Sans GB"
      style: W3
    - family: "Apple SD Gothic Neo"
    - family: "PingFang HK"
    - family: "Kohinoor Bangla"
    - family: "Kohinoor Devanagari"
    - family: "Gujarati Sangam MN"
    - family: "Gurmukhi MN"
    - family: "Kannada Sangam MN"
    - family: "Khmer Sangam MN"
    - family: "Lao Sangam MN"
    - family: "Malayalam Sangam MN"
    - family: "Myanmar Sangam MN"
    - family: "Oriya Sangam MN"
    - family: "Sinhala Sangam MN"
    - family: "Tamil Sangam MN"
    - family: "Kohinoor Telugu"
    - family: "Mshtakan"
    - family: "Euphemia UCAS"
    - family: "Plantagenet Cherokee"
    - family: "Apple Color Emoji"
```

**Comment by mscharley** [comment]
That is a long list. If this is a default can we also get another way to input known or discoverable lists like this without having to find that entire list ourselves?

**Comment by mscharley** [comment]
Perhaps 

```
# Font configuration (changes require restart)
font:
  # The normal (roman) font face to use.
  normal:
    - family: "Menlo" # should be "Menlo" or something on macOS.
      style: Regular  # Style can be specified to pick a specific face.
    - family: "Monaco"
    - preset: "Darwin fallbacks" 
```

**Comment by algesten** [comment]
Further looking into `glyph_index`, to get higher unicode numbers working, there are a couple of things to do.

The conversion from `char` to `UniChar` must go via UTF-16. I believe [this row](https://github.com/jwilm/alacritty/blob/32248471667dbdea33fca5e12fa0cb8a43482106/font/src/darwin/mod.rs#L446) only works for low code points where UTF-8/16 doesn't differ:

        let chars = [character as UniChar];

`UniChar` is a `c_ushort` which in turn is a `u16`. However a `char` can be anything up to 4 bytes.

[For `ᚦ`](https://unicode-table.com/en/16A6/) we should already be safe.

But for [`😉`](https://unicode-table.com/en/1F609/), the resulting UTF-16 would require ["surrogate pairs"](https://en.wikipedia.org/wiki/UTF-16#U.2B10000_to_U.2B10FFFF). For `CTFontGetGlyphsForCharacters` apple source code says

>    @param      characters
                An array of characters (UTF-16 code units). Non-BMP characters must be encoded as surrogate pairs.
>
>    @param      glyphs
>                A pointer to a buffer to receive the glyphs. Glyphs for non-BMP characters are sparse: the first glyph corresponds to the full character and the second glyph will be 0.

I.e. for these pairs we provide two `UniChar` in and will get one glyph and a 0 out.

Next up is figuring out how to do UTF-16 encoding in rust.

**Comment by algesten** [comment]
<img width="372" alt="screen shot 2017-05-15 at 16 09 55" src="https://cloud.githubusercontent.com/assets/227204/26061732/03f0885a-3989-11e7-96d2-4bf348330d85.png">

Great joy, and happiness :D

**Comment by jwilm** [comment]
@algesten 

I am very appreciative of the work you're doing for macOS font fallback. This is one of the biggest missing features on that platform, and also one of the most important. Thank you.

Few thoughts on design-

Instead of shipping a hard-coded cascade list for macOS, I would suggest we add a binding for `CTFontCopyDefaultCascadeListForLanguages`. Supporting a fallback list in the config is still a great idea; this allows prioritizing non-standard fonts like powerline variants and icon fonts like font-awesome.

For FreeType/fontconfig systems, we should dynamically build a fallback list whenever we perform a successful glyph fallback. Actually, it may be worthwhile to build dynamic lists on macOS as well so we don't continually search fonts that aren't being used.

**Comment by algesten** [comment]
@jwilm thanks for encouraging words! I just got a [working version](https://github.com/algesten/alacritty/commit/3ebeaa06baa1a8138b61b9d3d51ad426be9c9d26), but I need your input to take it all the way.

`CTFontCopyDefaultCascadeListForLanguages` is not in upstream `core-text-rs`, I will fork and PR the maintainers to get it in. I'll get going on that.

I've done it so that on startup, we try to load all the fonts in the list – any that fail to load is just removed and not attempted when rendering a glyph.

Not entirely clear what the difference will be for FreeType/fontconfig, but I was hoping a similar approach to macOS would work (will look at this next).

Questions:
* I get a feeling you're thinking dynamic loading, but currently that's not how I've done it. Do you think there's a problem loading all fonts on startup and that we ought to lazy-load? This would require some bigger changes in the interactions between `src/renderer/mod.rs` and `font/src/lib.rs`, which at this point are fairly minimal.
* For macOS, do you want to always add on `CTFontCopyDefaultCascadeListForLanguages` at the end of what the user provides in `alacritty.yml`?

**Comment by algesten** [comment]
For macOS it now always appends the [system fallback lists for `en`](b895cac66dc2b0325c0eece53036d203f587beef).

**Comment by algesten** [comment]
Hm. Looking at the freetype code I understand. It's quite a lot different from macOS. On macOS the font loading is done already on `load_font`, but on freetype it's done first on get_glyph().

**Comment by jwilm** [comment]
> CTFontCopyDefaultCascadeListForLanguages is not in upstream core-text-rs, I will fork and PR the maintainers to get it in. I'll get going on that.

👍 

> I get a feeling you're thinking dynamic loading, but currently that's not how I've done it. Do you think there's a problem loading all fonts on startup and that we ought to lazy-load? This would require some bigger changes in the interactions between src/renderer/mod.rs and font/src/lib.rs, which at this point are fairly minimal.

Seems fine to just load all fonts on startup especially if it's easier.

> For macOS, do you want to always add on CTFontCopyDefaultCascadeListForLanguages at the end of what the user provides in alacritty.yml?

Yes. Did you have some alternative in mind? I think this makes the most sense... always check user defined fallbacks before proceeding to system fallbacks.

> For macOS, do you want to always add on CTFontCopyDefaultCascadeListForLanguages at the end of what the user provides in alacritty.yml?

They were implemented independently at different times. The only real similarity is they share a public API.

To support user-provided fallbacks in FreeType and fontconfig, we need to build our own fallback list that's checked before the system fallback lists. With FreeType and fontconfig, you can just think of it like fontconfig is managing the cascading for us; we just provide it with a utf-32 codepoint and it provides a filtered list of fonts (or just one, depending on what API it's queried with).

The relevant code is in _ft/mod.rs_:

```rust
        let mut charset = fc::CharSet::new();
        charset.add(glyph);
        let mut pattern = fc::Pattern::new();
        pattern.add_charset(&charset);

        let config = fc::Config::get_current();
        match fc::font_match(config, &mut pattern) {
```

We use the default Config here to search all fonts. For user-provided lists, we should store our own config to be checked first.

I'm a bit disappointed that we need to support this at all for fontconfig. In theory, it's completely unnecessary. fontconfig can be configured at the user/system level to prioritize fonts for fallback.

**Comment by algesten** [comment]
Ok. Cool. I've started refactoring to get more similarity and code reuse between the macOS and freetype versions. Even though it will take longer to get this out the door, I really do think it's better in the long run.

**Comment by algesten** [comment]
My effort is now tracked in #575. Almost done.

**Comment by algesten** [comment]
Ok. I consider it done. Would appreciate if someone Linux based tried it out (I'm running in VWMare).

Feedback welcome.

**Comment by cpick** [comment]
I'm excited about this working properly, lots of great work @algesten!

I gave the changes from #575 a quick test on Linux (Ubuntu 16.04, xenial) and after I updated my `alacritty.yml` so that my listed font families were a list they worked as advertised.  Specifically, my vim powerline looked correct when I set:

```yml
font:
  normal:
    - family: monospace
    - family: PowerlineSymbols
```

**However**, I should *not* have to do this in my alacritty config.  My user's fontconfig is already setup to fall back to PowerlineSymbols when required to render a character.

My cpick:font-fallback branch from #456 accomplishes this by having `FreeTypeRasterizer`'s `get_matching_face()`, `get_specific_face()`, and `get_face_descs()` return `FaceDescs` instead of a `Face`.  This allows `FreeTypeRasterizer::get_glyph()` to use `face_desc.charset.contains()` to search for the right font from the user's fontconfig.

**Comment by NickeZ** [comment]
I've compiled it but I'm not sure how to test it. I tried with emojis but I think my font already has them..

**Comment by jwilm** [comment]
@cpick thanks for the report

It's important that system/user configuration for fontconfig continues to be respected. This means that for fontconfig, we should maintain two sets: 1, the custom set defined by config and 2, the system font set. Fallbacks can be checked in that order (alacritty config, system/user config).

I still haven't had a chance to review the PR associated with this, but it may be that a different abstraction is needed for checking for fallbacks if such behavior can't be supported with the new design.

**Comment by algesten** [comment]
@cpick thanks! @jwilm I think what @cpick is saying is that there is an _event better_ way of doing freetype/fontconfig that currently lives in #456. My PR works off master, so hasn't got those changes, However unless I screwed up, it should work like current master.

The intention is that like @jwilm describes, regardless of macOS/freetype.

1. config (alacritty.yml)
2. fallbacks

Happy to look at incorporating #456 in my patch if everyone thinks that's better.

**Comment by cpick** [comment]
@algesten that's exactly right.

I think #575 is a great step forward in functionality (I didn't review the code itself): allowing users to configure custom font-fallbacks within alacritty works as advertised on Linux.

Eventually/additionally getting fontconfig's fallback working is a good/complimentary next step, but I wouldn't think it would be a blocker to merging this improvement.

**Comment by maximbaz** [comment]
Just tried on my linux, seems to be working good :+1: 

I've got a question, what happens (or should happen) to the `offset` and `glyph_offset` configs? They are global now, but shouldn't they become per-family configurations? I always thought of them as a patch for a specific font, so if I have `monospace` and `PowerlineSymbols` defined, I might need to adjust the glyph offset only for the `PowerlineSymbols` font.... or not?

**Comment by jwilm** [comment]
@maximbaz `offset` really is more of a global option since it influences cell size/placement. I agree that `glyph_offset` should be configurable per font. Kind of out of scope for this PR, though.

**Comment by algesten** [comment]
@jwilm there was one thing I forgot to ask you. I removed [`cg_font`](https://github.com/jwilm/alacritty/blob/master/font/src/darwin/mod.rs#L300), because nothing seemed to read the field. Did I miss something? Or was it for the future?

**Comment by sodiumjoe** [comment]
hey, super excited about this work, I tried @algesten's branch, but I can't get emoji to work. I used the font list [here](https://github.com/jwilm/alacritty/issues/45#issuecomment-301343001), but I still get blank spaces for emoji.

**Comment by algesten** [comment]
Since the changes that adds on apple fallbacks, we don't need to the full list (shouldn't hurt either since it removes duplicates):

```
# Font configuration (changes require restart)
font:
  # The normal (roman) font face to use.
  normal:
    - family: Menlo
    - family: Apple Symbols

  # The bold font face
  bold:
    - family: Menlo
      style: Bold
    - family: Apple Symbols

  # The italic font face
  italic:
    - family: Menlo
      style: Italic
    - family: Apple Symbols
```

@sodiumjoe Just to double check, you're on macOS or Linux?

**Comment by sodiumjoe** [comment]
That works, thanks! This is awesome! (macOS Sierra 10.12.4, btw)

**Comment by polyzen** [comment]
@algesten, are you still planning to add Linux font fallbacks? Assumed it was going to follow shortly after the rest of #575 being merged, but I was sadly mistaken. :p

**Comment by sodiumjoe** [comment]
I believe this is closed by https://github.com/jwilm/alacritty/pull/956

**Comment by mathstuf** [comment]
@sodiumjoe That appears to be macos-only. It seems Linux (freetype/fontconfig) support is still not there.

**Comment by chrisduerr** [comment]
Linux font fallback should work. It just uses the first font it can find which supports the specified glyph. Both on macos and linux it's not yet possible to specify specific fallback fonts, but that might be fixed after #957.

**Comment by sharpner** [comment]
For me this behaves not quite as I want it to be, do I have to enable a new config setting or something? 

I have been using this patch https://github.com/algesten/alacritty/commit/1ecd26ef0467b367d8c59522ca0c60004f7a6436.patch

up till now, and the results are different. So with the patch:

![old-patch](https://user-images.githubusercontent.com/574362/35849456-e58e2be0-0b21-11e8-9477-baebbb8cd53e.png)


And with the newest unpatched alacritty master (2466f81d5e0cb56dd00231f63f2dbcef1840196a):
![new](https://user-images.githubusercontent.com/574362/35849388-b8509118-0b21-11e8-9f0d-044ef00db064.png)

The icon ➤ is not showing, and also the ✗ is missing.

**Comment by sodiumjoe** [comment]
I believe this is a regression https://github.com/jwilm/alacritty/issues/1086

**Comment by jwilm** [comment]
Closing this since fallback has been implemented on master for a long time.

**Comment by polyzen** [comment]
@sharpner @sodiumjoe is the regression fixed for you?

**Comment by sodiumjoe** [comment]
fixed for me, yes

**Comment by sharpner** [comment]
for me too! thx
